src/machine/stack.rs

```diff
@@ -86,7 +94,8 @@ impl Index<usize> for Stack {
     #[inline]
     fn index(&self, index: usize) -> &Self::Output {
         unsafe {
-            let ptr = self.buf.base as usize + index;
+            // TODO: implement some mechanism to verify soundness
+            let ptr = self.buf.get_unchecked(index);
```

Comment:
Can't this just do this?

```suggestion
            let ptr = self.buf.get(index).unwrap();
```

Reply:
It can, though I am worried about the loss in performance. I haven't spun up any benchmarks on my end yet

Reply:
Correctness is more important than performance for us in general. But seeing as this is very low level and probably on the hot path everywhere, it may make sense to just mark these currently unsound functions as unsafe instead.

Reply:
My intuition is that the performance impact wouldn't be that great actually, but we would need to measure it to be sure.

Reply:
Also, a robust and correct core that allows automated safety checks may even more readily enable performance&nbsp;gains from completely different&nbsp;sources in the future, such as JIT&nbsp;compilation. As long as performance remains somewhat acceptable, I would favor everything that helps to make the engine more&nbsp;correct.

Reply:
From quickly running the benchmarks, the impact of a simple assertions like this one is too small to be measurable, so let's err on the side of safety :)